hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -0,0 +1,218 @@
+import com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.SqlCloseCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlExecuteCodec;
+import com.hazelcast.client.impl.protocol.codec.SqlFetchCodec;
+import com.hazelcast.client.impl.spi.impl.ClientInvocation;
+import com.hazelcast.client.impl.spi.impl.ClientInvocationFuture;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.SqlException;
+import com.hazelcast.sql.SqlQuery;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.row.Row;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Client-side implementation of SQL service.
+ */
+public class SqlClientService implements SqlService {
+
+    private static final ClientMessageDecoder<SqlExecuteResponse> EXECUTE_DECODER = clientMessage -> {
+        SqlExecuteCodec.ResponseParameters response = SqlExecuteCodec.decodeResponse(clientMessage);
+
+        return new SqlExecuteResponse(response.queryId, response.rowMetadata, response.rowPage, response.error);
+    };
+
+    private static final ClientMessageDecoder<SqlFetchResponse> FETCH_DECODER = clientMessage -> {
+        SqlFetchCodec.ResponseParameters response = SqlFetchCodec.decodeResponse(clientMessage);
+
+        return new SqlFetchResponse(response.rowPage, response.error);
+    };
+
+    private static final ClientMessageDecoder<Void> CLOSE_DECODER = clientMessage -> {
+        SqlCloseCodec.decodeResponse(clientMessage);
+
+        return null;
+    };
+
+    private final HazelcastClientInstanceImpl client;
+
+    public SqlClientService(HazelcastClientInstanceImpl client) {
+        this.client = client;
+    }
+
+    @Nonnull
+    @Override
+    public SqlResult query(@Nonnull SqlQuery query) {
+        Connection connection = client.getConnectionManager().getRandomConnection(true);
+
+        if (connection == null) {
+            throw rethrow(QueryException.error(
+                SqlErrorCode.CONNECTION_PROBLEM,
+                "Client must be connected to at least one data member to execute SQL queries"
+            ));
+        }
+
+        try {
+            List<Object> params = query.getParameters();
+            List<Data> params0;
+
+            if (!params.isEmpty()) {
+                params0 = new ArrayList<>(params.size());
+
+                for (Object param : params) {
+                    params0.add(serializeParameter(param));
+                }
+            } else {
+                params0 = null;
```

Comment:
It would be much easier to understand if we represent the empty list with an actually empty list rather than `null`. 
I have checked the `SqlExecuteMessageTask` code. It would simplify the code there also. 
Instead of 
```
 if (parameters.parameters != null && !parameters.parameters.isEmpty()) {
                for (Data param : parameters.parameters) {
                    query.addParameter(serializationService.toObject(param));
                }
            }
```
the following will suffice
```
                for (Data param : parameters.parameters) {
                    query.addParameter(serializationService.toObject(param));
                }      
```

Reply:
Fixed, now parameters are non-null.